# TCSS 588 In-class Exercise: Correlation and linear regression

These exercises will be used to illustrate how to perform data analyses using R.
Alternatively, you can use R or Rstudio. 
I assume you will learn basic R commands on your own.

Credit: The data and ideas behind these exercises and homeworks are from the NIH LINCS DCIC Crowdsourcing Portal and Ma’ayan Lab @ Mt Sinai, New York.
http://www.maayanlab.net/crowdsourcing/megatask1.php

The overarching goal is to predict adverse drug reactions. But we will take this task step by step. You will learn more about the meanings of these data in future lectures. For illustrative purposes in today’s class, just treat these data as numbers in a matrix.

## 1. Download and read in data (same as lecture 2)
Download the data “gene_expression_n438x978.txt” from canvas to your computer.  I have removed the first character (#) and an ‘ from the input file to make R happy.

In [11]:
my.data <- read.table ("gene_expression_n438x978.txt", sep="\t", header=T)
dimnames(my.data)[[1]] <- my.data[,1]
my.data <- my.data[, -1]
dim (my.data)
my.data[1:5, 1:5]

[1] 438 978

PSME1        ATF1         RHEB         FOXO3      
ACETAZOLAMIDE       -0.015159099 -0.031470528 -0.004733488  0.02591061
IRBESARTAN          -0.026811981  0.012151979 -0.025550148 -0.02401181
IPRATROPIUM BROMIDE  0.001017958 -0.008650622 -0.018128698 -0.02079971
EFAVIRENZ           -0.004398264  0.055387992  0.004658520  0.01380732
THIAMINE             0.001838965 -0.018079188 -0.011855532 -0.03705033
                    RHOA         
ACETAZOLAMIDE        0.0056296773
IRBESARTAN          -0.0106717396
IPRATROPIUM BROMIDE -0.0002722781
EFAVIRENZ           -0.0340697348
THIAMINE            -0.0133954959

## 2. Correlation (same as lecture)

In [ ]:
cor (my.data[,1], my.data[, 2])

## 3. Pairwis correlation

_Your task 1:_ What is the maximum correlation coefficient among all possible pairs of genes?  How about the minimum correlation?  

_Q:_ Can you get away from doing a double nested loop in R?
Discuss with your neighbors.

Note that the total number of possible gene pairs = 978*977/2 = 477,753

_Your task 2:_ write a R script. For computationally intensive tasks, you might want to consider running your R code in batch mode.  https://www.r-bloggers.com/batch-processing-vs-interactive-sessions/

Hint: There are many ways to do this. You can use a loop or you might want to check out
http://www.r-bloggers.com/using-apply-sapply-lapply-in-r/


In [ ]:
curr.max <- -1
curr.min <-1
for (i in 1:ncol(my.data)) {
    for(j in 1:nrows(my.data)) {
    ret.cor <- cor (my.data[,i], my.data[,j])
    curr.max <- max(c(curr.max, ret.cor))
    curr.min <- min(c(curr.min, ret.cor))
}
}

In [12]:
help (lm)

In [ ]:
lmfit1 <- lm(my.data$CDK)